### PR TITLE
formulae_dependents: fix linkage test issues

### LINKED ISSUE #611: Linkage tests are passing when they should fail
`clazy` passed CI in Homebrew/homebrew-core#77005, but then failed the linkage check in [this CI run](https://github.com/Homebrew/homebrew-core/actions/runs/848283367) (Homebrew/homebrew-core#77434).

It was then given a `revision` bump to rebuild it, and I thought that shook off the linkage (as CI passed after the `revision` bump), but the new bottle still shows broken linkage:
```
❯ brew linkage clazy
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libncurses.5.4.dylib
  /usr/lib/libxml2.2.dylib
  /usr/lib/libz.1.dylib
Variable-referenced libraries:
  @rpath/ClazyPlugin.dylib
Broken dependencies:
  /usr/local/opt/llvm@11/lib/libc++.1.dylib (llvm@11)
Dependencies with no linkage:
  qt
```

This seems to be the step that causes it: https://github.com/Homebrew/homebrew-core/runs/2597888223#step:7:768

### PR DESCRIPTION
**Pull request by carlocab**
We see two kinds of spurious linkage test issues in CI:
1. tests failing when they should succeed (e.g. Homebrew/homebrew-core#85367)
2. tests succeeding when they should fail (#611).

This should fix both problems by either uninstalling dependents that
aren't necessarily around at runtime, or linking dependencies that are.

Closes #611.

### CODE REVIEW

**Comment by carlocab** [comment]
Actually, this probably doesn't close #611. Will need to think about this some more.

File: lib/tests/formulae_dependents.rb [hunk 1]
```diff
@@ -166,6 +166,23 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
           test "brew", "link", dependent.full_name
         end
         test "brew", "install", "--only-dependencies", dependent.full_name
+
+        dependent.deps.each do |dependency|
+          dependency_f = dependency.to_formula
+
+          # Uninstall build- or test-only dependencies to avoid `brew linkage --test`
```
Comment (MikeMcQuaid) [inline lib/tests/formulae_dependents.rb:173]:
```suggestion
          # Uninstall build-only or test-only dependencies to avoid `brew linkage --test`
```

Comment (MikeMcQuaid) [reply lib/tests/formulae_dependents.rb:173]: Also, we already do uninstall this for non-dependents testing, perhaps we could/should use the logic (or adjust the ordering?)

File: lib/tests/formulae_dependents.rb [hunk 2]
```diff
@@ -166,6 +166,23 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
...
+          # mistakenly succeeding. Homebrew/homebrew-test-bot#611. Link them otherwise
+          # to avoid spurious CI failures.
+          subcommand = if (dependency.build? || dependency.test?) && dependency_f.any_version_installed?
+            ["uninstall", "--ignore-dependencies"]
```
Comment (MikeMcQuaid) [inline lib/tests/formulae_dependents.rb:177]: Why do we ignore dependencies here, does this not mean we risk uninstalling stuff actually required?

File: lib/tests/formulae_dependents.rb [hunk 3]
```diff
@@ -166,6 +166,23 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
...
+          elsif !dependency_f.keg_only? && !dependency_f.linked_keg.exist?
+            unlink_conflicts dependency_f
+            ["link"]
```
Comment (MikeMcQuaid) [inline lib/tests/formulae_dependents.rb:180]: So, this doesn't have to be in this PR but given `brew install` doesn't actually verify all dependents are linked (right,  correct me if I'm wrong?) does this not mean we can hide legitimate failures.

I think ideally we'd actually not require any dependency to be linked for the `brew linkage` to work (but that may be too involved to do for now).

**Review by MikeMcQuaid — APPROVED** [review]
Despite comments: fine with merging this whenever you're happy with it.

File: lib/tests/formulae_dependents.rb (hunk 1 above)
Comment (carlocab) [reply lib/tests/formulae_dependents.rb:173]: Yea, this step smells reusable to me. I just haven't worked out the right way to refactor things. I'll take a closer look.

File: lib/tests/formulae_dependents.rb (hunk 2 above)
Comment (carlocab) [reply lib/tests/formulae_dependents.rb:177]: We ignore dependencies because something else installed might need the build/test-only dependency we want to uninstall.

One thing I need to think more about here is the case where the build/test-only dependency is a modified formula: I suspect uninstalling those will require special reinstall logic which we currently don't have.

File: lib/tests/formulae_dependents.rb (hunk 3 above)
Comment (carlocab) [reply lib/tests/formulae_dependents.rb:180]: > So, this doesn't have to be in this PR but given `brew install` doesn't actually verify all dependents are linked (right, correct me if I'm wrong?) does this not mean we can hide legitimate failures.

Yes, I believe `brew install foo` doesn't require that everything that `foo` depends on is linked. I'm not sure I follow how this hides legitimate failures, though.

> I think ideally we'd actually not require any dependency to be linked for the `brew linkage` to work (but that may be too involved to do for now).

We actually don't require linked dependencies for `brew linkage` to work on macOS. In fact, barring the odd exception (e.g. GCC), `brew unlink foo` should never break anything on macOS. This is not true on Linux, because we don't use `opt`-paths when encoding `RPATH`s for non-keg-only dependencies, so doing `brew link` is only needed to fix Linux CI.

Fixing up `RPATH`s on Linux is probably not too involved, though I'm a little wary of changing things there since I'm much less familiar with Linux.

File: lib/tests/formulae_dependents.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply lib/tests/formulae_dependents.rb:177]: > We ignore dependencies because something else installed might need the build/test-only dependency we want to uninstall.

Do we install it again later if it's needed?

> One thing I need to think more about here is the case where the build/test-only dependency is a modified formula: I suspect uninstalling those will require special reinstall logic which we currently don't have.

I think we have that logic just not for dependants (uses the bottles) or we just don't install them in that case. The existing logic I linked in the issue should help with these edge cases.

File: lib/tests/formulae_dependents.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply lib/tests/formulae_dependents.rb:180]: > Yes, I believe `brew install foo` doesn't require that everything that `foo` depends on is linked. I'm not sure I follow how this hides legitimate failures, though.

I guess it depends if you consider `brew unlink $dependency; brew install foo; foo` failing with a linkage area to be a legitimate failure. I kinda do?

> We actually don't require linked dependencies for `brew linkage` to work on macOS. In fact, barring the odd exception (e.g. GCC), `brew unlink foo` should never break anything on macOS. This is not true on Linux, because we don't use `opt`-paths when encoding `RPATH`s for non-keg-only dependencies, so doing `brew link` is only needed to fix Linux CI.

Gotcha 👍🏻 I care less about Linux 🤣 

> Fixing up `RPATH`s on Linux is probably not too involved, though I'm a little wary of changing things there since I'm much less familiar with Linux.

Yeh, I'd be fine with this being mainly to handle the Linux case (although maybe worth gating the `link` to Linux only in that case?) but I think it'd be great to have a PR to do this `opt` change on Linux, too.
